presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressions.java

```diff
@@ -0,0 +1,280 @@
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.google.common.base.Predicates.not;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public final class RowExpressions
+{
+    public static final ConstantExpression TRUE = new ConstantExpression(true, BOOLEAN);
+    public static final ConstantExpression FALSE = new ConstantExpression(false, BOOLEAN);
+
+    private final DeterminismEvaluator determinismEvaluator;
+
+    public RowExpressions(FunctionManager functionManager)
+    {
+        this.determinismEvaluator = new DeterminismEvaluator(requireNonNull(functionManager, "functionManager is null"));
+    }
+
+    public static List<RowExpression> extractConjuncts(RowExpression expression)
+    {
+        return extractPredicates(AND, expression);
+    }
+
+    public static List<RowExpression> extractDisjuncts(RowExpression expression)
+    {
+        return extractPredicates(OR, expression);
+    }
+
+    public static List<RowExpression> extractPredicates(RowExpression expression)
+    {
+        if (expression instanceof SpecialFormExpression) {
+            Form form = ((SpecialFormExpression) expression).getForm();
+            if (form == AND || form == OR) {
+                return extractPredicates(form, expression);
+            }
+        }
+        return ImmutableList.of(expression);
+    }
+
+    public static List<RowExpression> extractPredicates(Form form, RowExpression expression)
+    {
+        if (expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm() == form) {
+            SpecialFormExpression specialFormExpression = (SpecialFormExpression) expression;
+            verify(specialFormExpression.getArguments().size() == 2, "logical binary expression requires exactly 2 operands");
+            return ImmutableList.<RowExpression>builder()
+                    .addAll(extractPredicates(form, specialFormExpression.getArguments().get(0)))
+                    .addAll(extractPredicates(form, specialFormExpression.getArguments().get(1)))
+                    .build();
+        }
+
+        return ImmutableList.of(expression);
+    }
+
+    public static RowExpression and(RowExpression... expressions)
+    {
+        return and(Arrays.asList(expressions));
+    }
+
+    public static RowExpression and(Collection<RowExpression> expressions)
+    {
+        return binaryExpression(AND, expressions);
+    }
+
+    public static RowExpression or(RowExpression... expressions)
+    {
+        return or(Arrays.asList(expressions));
+    }
+
+    public static RowExpression or(Collection<RowExpression> expressions)
+    {
+        return binaryExpression(OR, expressions);
+    }
+
+    public static RowExpression binaryExpression(Form form, Collection<RowExpression> expressions)
```

Comment:
This function unflatten the LogicalExpression to binary form. Maybe buildLogicalBinaryTree?

Reply:
This function is only used in this class so maybe consider making it private?

Reply:
RowExpression for logical operation supports either flatten form and unflatten form, maybe in the future we don't need to flatten/unflatten if so much?

Reply:
I would still keep the interface names and accessibility sync with `ExpressionUtils`

Reply:
Yes, fine with me. The original name is ambiguous so maybe a comment on what it does (I know there are some below but sometimes if reviewing from web page, you can only see the function name)? binaryExpression gives impression that it is operation on binary data ...